### PR TITLE
support a behaviour like 'with' when using roles

### PR DESCRIPTION
**Pull request by reneeb**
### Summary
Currently Mojo::Base's with_roles creates a new class, but we often need to apply the roles
to the package. So this is basically the behavior of Moose's `with` and/or `Role::Tiny::With`

### Motivation
We do not want to apply the roles when we create a new object. We want the role behavior included in the classes.

### CODE REVIEW

**Comment by jhthorsen** [comment]
I will not be in favor of this PR, since it's way too much typing to use it. I think I would rather have something like:

```
use Mojo::Base 'Some::Class', -with => ['Some::Role'];
use Mojo::Base -with => 'Some::Role'; # "-base" implied?
```

**Comment by Grinnz** [comment]
This is already provided by Role::Tiny::With.

**Comment by reneeb** [comment]
I like @jhthorsen  suggestion.

@Grinnz: It feels weird to use Role::Tiny::With to use one Role::Tiny feature while other Role::Tiny features can be used with a Mojo::Base method.

**Comment by Grinnz** [comment]
Not sure what you mean. All of the Mojo::Base role features are just wrappers to use Role::Tiny properly. There is no wrapper required here.

**Comment by reneeb** [comment]
But you suggested the use of Role::Tiny::With to have the 'with' like in Moose...

**Comment by Grinnz** [comment]
Yes, this is the appropriate way to compose a role statically into a class.

**Comment by reneeb** [comment]
To be more precise, to use Role::Tiny's `apply_roles_to_object` and `create_class_with_roles` can be used with `with_roles`, but not `apply_roles_to_package`.

**Comment by jberger** [comment]
I tend to agree with @Grinnz and @jhthorsen here. You can already accomplish this with the Role::Tiny standard module Role::Tiny::With. If we wanted something in core, I'd vote for adding a -with flag to be consistent with the existing Mojo::Base mechanisms

**Comment by reneeb** [comment]
I've updated the PR to use the "-with" flag that @jhthorsen proposed...

File: lib/Mojo/Base.pm [hunk 1]
```diff
@@ -385,10 +402,6 @@ new class. For roles following the naming scheme C<MyClass::Role::RoleName> you
 can use the shorthand C<+RoleName>. Note that role support depends on
 L<Role::Tiny> (2.000001+).
 
-  # Create a new class with the role "SubClass::Role::Foo" and instantiate it
-  my $new_class = SubClass->with_roles('+Foo');
-  my $object    = $new_class->new;
```
Comment (jhthorsen) [inline lib/Mojo/Base.pm]: Why are these lines removed?

Comment (reneeb) [reply lib/Mojo/Base.pm]: My mistake...

**Comment by reneeb** [comment]
@jhthorsen I'll have a look at it tomorrow...

File: lib/Mojo/Base.pm [hunk 2]
```diff
@@ -127,6 +127,18 @@ sub import {
...
     }
 
+    # apply role to package
+    elsif ($flag eq '-with') {
+      Carp::croak 'Role::Tiny 2.000001+ is required for roles' unless ROLES;
+
+      my $role_list = shift @flags;
+      my @roles = map {
+        /^\+(.+)$/ ? "${caller}::Role::$1" : $_
+      } (ref $role_list ? @{$role_list} : $role_list);
```
Comment (kraih) [inline lib/Mojo/Base.pm]: I bet this block could be much shorter.

File: t/mojo/roles.t [hunk 3]
```diff
@@ -46,6 +46,25 @@ use Mojo::Base -role;
 
 sub hello {'hello mojo!'}
 
+package Mojo::RoleListApplied::Role::LOUD;
```
Comment (kraih) [inline t/mojo/roles.t]: We include the word `Test` in test class names.

File: t/mojo/roles.t [hunk 4]
```diff
@@ -124,5 +143,16 @@ is $dom->hello, 'hello mojo!', 'right result';
 my $file = Mojo::File->with_roles('Mojo::RoleTest::Hello')->new;
 is $file->hello, 'hello mojo!', 'right result';
 
+# role applied to package, do not create a new class
```
Comment (kraih) [inline t/mojo/roles.t]: Inconsistently formatted comment.

File: t/mojo/roles.t [hunk 5]
```diff
@@ -124,5 +143,16 @@ is $dom->hello, 'hello mojo!', 'right result';
...
+my $list_applied = Mojo::RoleListApplied->new;
+is $list_applied->hello, 'hello mojo!';
+is $list_applied->yell, 'HEY!';
+
```
Comment (kraih) [inline t/mojo/roles.t]: Extra newline.

File: t/mojo/roles.t [hunk 6]
```diff
@@ -124,5 +143,16 @@ is $dom->hello, 'hello mojo!', 'right result';
...
+my $role_applied = Mojo::RoleApplied->new;
+is $role_applied->yell, 'HEY!';
+
```
Comment (kraih) [inline t/mojo/roles.t]: Extra newline.

File: lib/Mojo/Base.pm [hunk 7]
```diff
@@ -260,6 +272,11 @@ enable support for L<subroutine signatures|perlsub/"Signatures">.
 This will also disable experimental warnings on versions of Perl where this
 feature was still experimental.
 
+If you use Roles, you can apply them to your class with C<-with>:
+
+  use Mojo::Base -with => ['Some::Role','+Logger'];
+  use Mojo::Base -with => 'Some::Role';
```
Comment (kraih) [inline lib/Mojo/Base.pm]: Shouldn't there be another flag?

File: lib/Mojo/Base.pm [hunk 8]
```diff
@@ -260,6 +272,11 @@ enable support for L<subroutine signatures|perlsub/"Signatures">.
 This will also disable experimental warnings on versions of Perl where this
 feature was still experimental.
 
+If you use Roles, you can apply them to your class with C<-with>:
```
Comment (kraih) [inline lib/Mojo/Base.pm]: Inconsistent formatting.

Comment (kraih) [reply lib/Mojo/Base.pm]: And description of the feature.

File: lib/Mojo/Base.pm (hunk 2 above)
Comment (jhthorsen) [reply lib/Mojo/Base.pm]: Here's my idea:

```perl
my $roles  = shift @flags;
my $prefix = "${caller}::Role::";
Role::Tiny->apply_roles_to_package($caller,
  map { /^\+(.+)/ ? "$prefix$1" : $_ } ref $roles ? @$roles : $roles);
```

File: lib/Mojo/Base.pm [hunk 9]
```diff
@@ -260,6 +272,11 @@ enable support for L<subroutine signatures|perlsub/"Signatures">.
...
+
+  use Mojo::Base -with => ['Some::Role','+Logger'];
```
Comment (jhthorsen) [inline lib/Mojo/Base.pm]: Missing whitespace inside the array-ref.

**Comment by kraih** [comment]
As we've recently learned from the `-async` flag, you can't use commonly used function names for flags. So using `-with` here is impossible.

**Comment by kraih** [comment]
I'm rejecting this PR without a vote for the reasons mentioned above. Too dangerous.

**Comment by shadowcat-mst** [comment]
package MyApp::Base;
    
    use Mojo::Base -strict;
    use Role::Tiny::With ();
    use Import::Into;
    
    sub import {
        shift;
        Mojo::Base->import::into(1, @_);
        Role::Tiny::With->import::into(1);
    }
    
    1;
